File: src/weather.c
/*
 * weather.c - NOAA weather API client
 *
 * Two-step API:
 *   1. GET https://api.weather.gov/points/{lat},{lon}
 *      -> extract properties.forecastHourly URL
 *   2. GET that URL
 *      -> extract first period's shortForecast, temperature, isDaytime
 *
 * Cloud cover is derived from forecast keyword heuristic (no direct cloud %
 * in the hourly forecast -- NOAA provides probabilityOfPrecipitation instead).
 */

#include "weather.h"
#include "json.h"

#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>

/* Dynamic response buffer for libcurl */
typedef struct {
    char   *data;
    size_t  size;
    size_t  cap;
} response_buf_t;

static size_t write_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
{
    response_buf_t *buf = userdata;
    size_t total = size * nmemb;

    if (buf->size + total + 1 > buf->cap) {
        size_t new_cap = (buf->cap ? buf->cap * 2 : 4096);
        while (new_cap < buf->size + total + 1) new_cap *= 2;
        char *new_data = realloc(buf->data, new_cap);
        if (!new_data) return 0;
        buf->data = new_data;
        buf->cap = new_cap;
    }

    memcpy(buf->data + buf->size, ptr, total);
    buf->size += total;
    buf->data[buf->size] = '\0';
    return total;
}

static bool http_get(const char *url, response_buf_t *resp)
{
    resp->data = nullptr;
    resp->size = 0;
    resp->cap = 0;

    CURL *curl = curl_easy_init();
    if (!curl) return false;

    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "User-Agent: abraxas/1.0 (weather color temp daemon)");
    headers = curl_slist_append(headers, "Accept: application/geo+json");

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, resp);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);

    CURLcode res = curl_easy_perform(curl);

    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);

    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK || http_code != 200) {
        free(resp->data);
        resp->data = nullptr;
        resp->size = 0;
        return false;
    }

    return true;
}

/*
 * Cloud cover heuristic from NOAA forecast keywords.
 * Ordered by priority: more specific matches first.
 */
static int cloud_cover_from_forecast(const char *forecast)
{
    if (!forecast) return 0;

    /* Convert to lowercase for matching */
    char lower[128];
    size_t i = 0;
    for (; forecast[i] && i < sizeof(lower) - 1; i++)
        lower[i] = (char)(forecast[i] | 0x20); /* ASCII tolower */
    lower[i] = '\0';

    /* Precipitation always means heavy cloud */
    if (strstr(lower, "rain") || strstr(lower, "storm") ||
        strstr(lower, "snow") || strstr(lower, "drizzle") ||
        strstr(lower, "showers"))
        return 95;

    /* Overcast */
    if (strstr(lower, "overcast"))
        return 90;

    /* Mostly cloudy (before general "cloudy" check) */
    if (strstr(lower, "mostly cloudy"))
        return 75;

    /* Cloudy */
    if (strstr(lower, "cloudy"))
        return 90;

    /* Partly */
    if (strstr(lower, "partly"))
        return 50;

    /* Mostly sunny/clear (before general "sunny"/"clear") */
    if (strstr(lower, "mostly sunny") || strstr(lower, "mostly clear"))
        return 25;

    /* Sunny/clear */
    if (strstr(lower, "sunny") || strstr(lower, "clear"))
        return 10;

    return 0;
}

weather_data_t weather_fetch(double lat, double lon)
{
    weather_data_t wd = {
        .cloud_cover = 0,
        .forecast = "Unknown",
        .temperature = 0.0,
        .is_day = true,
        .fetched_at = time(nullptr),
        .has_error = true
    };

    /* Step 1: Get grid point */
    char url[256];
    snprintf(url, sizeof(url), "https://api.weather.gov/points/%.4f,%.4f", lat, lon);

    response_buf_t resp;
    if (!http_get(url, &resp)) return wd;

    json_value_t *root = json_parse(resp.data);
    free(resp.data);
    if (!root) return wd;

    const char *forecast_url = json_string(json_path(root, "properties.forecastHourly"));
    if (!forecast_url) {
        json_free(root);
        return wd;
    }

    /* Copy URL before freeing root */
    char hourly_url[512];
    snprintf(hourly_url, sizeof(hourly_url), "%s", forecast_url);
    json_free(root);

    /* Step 2: Get hourly forecast */
    if (!http_get(hourly_url, &resp)) return wd;

    root = json_parse(resp.data);
    free(resp.data);
    if (!root) return wd;

    const json_value_t *periods = json_path(root, "properties.periods");
    const json_value_t *period = json_at(periods, 0);

    if (!period) {
        json_free(root);
        return wd;
    }

    /* Extract forecast data */
    const char *short_forecast = json_string(json_get(period, "shortForecast"));
    if (short_forecast) {
        strncpy(wd.forecast, short_forecast, sizeof(wd.forecast) - 1);
        wd.forecast[sizeof(wd.forecast) - 1] = '\0';
    }

    const json_value_t *temp_val = json_get(period, "temperature");
    if (temp_val) wd.temperature = json_number(temp_val);

    const json_value_t *day_val = json_get(period, "isDaytime");
    if (day_val) wd.is_day = json_bool(day_val);

    /* Cloud cover from keyword heuristic */
    wd.cloud_cover = cloud_cover_from_forecast(wd.forecast);

    wd.has_error = false;
    json_free(root);
    return wd;
}

File: rust/src/weather.rs
//! NOAA weather API client.
//!
//! Two-step API:
//!   1. GET https://api.weather.gov/points/{lat},{lon}
//!      -> extract properties.forecastHourly URL
//!   2. GET that URL
//!      -> extract first period's shortForecast, temperature, isDaytime
//!
//! Uses curl(1) child process for HTTP -- zero TLS dependencies.
//! When compiled without the "noaa" feature, all functions are no-ops.

use crate::config::WeatherData;
use crate::now_epoch;

#[cfg(feature = "noaa")]
pub fn init() {}

#[cfg(feature = "noaa")]
pub fn cleanup() {}

#[cfg(feature = "noaa")]
pub fn fetch(lat: f64, lon: f64) -> WeatherData {
    match fetch_inner(lat, lon) {
        Ok(wd) => wd,
        Err(_) => WeatherData {
            cloud_cover: 0,
            forecast: "Unknown".to_string(),
            temperature: 0.0,
            is_day: true,
            fetched_at: now_epoch(),
            has_error: true,
        },
    }
}

#[cfg(feature = "noaa")]
fn http_get(url: &str) -> Result<String, Box<dyn std::error::Error>> {
    let output = std::process::Command::new("curl")
        .args([
            "-s", "-f", "-L", "--max-time", "5",
            "-H", "User-Agent: abraxas/6.1 (weather color temp daemon)",
            "-H", "Accept: application/geo+json",
            url,
        ])
        .output()?;

    if !output.status.success() {
        return Err(format!("curl exit {}", output.status).into());
    }

    String::from_utf8(output.stdout).map_err(|e| e.into())
}

#[cfg(feature = "noaa")]
fn fetch_inner(lat: f64, lon: f64) -> Result<WeatherData, Box<dyn std::error::Error>> {
    // Step 1: Get grid point
    let url = format!("https://api.weather.gov/points/{:.4},{:.4}", lat, lon);
    let body = http_get(&url)?;
    let resp: serde_json::Value = serde_json::from_str(&body)?;

    let forecast_url = resp["properties"]["forecastHourly"]
        .as_str()
        .ok_or("No forecastHourly URL")?
        .to_string();

    // Step 2: Get hourly forecast
    let body = http_get(&forecast_url)?;
    let resp: serde_json::Value = serde_json::from_str(&body)?;

    let period = &resp["properties"]["periods"][0];
    if period.is_null() {
        return Err("No forecast periods".into());
    }

    let short_forecast = period["shortForecast"]
        .as_str()
        .unwrap_or("Unknown")
        .to_string();
    let temperature = period["temperature"].as_f64().unwrap_or(0.0);
    let is_day = period["isDaytime"].as_bool().unwrap_or(true);

    let cloud_cover = cloud_cover_from_forecast(&short_forecast);

    Ok(WeatherData {
        cloud_cover,
        forecast: short_forecast,
        temperature,
        is_day,
        fetched_at: now_epoch(),
        has_error: false,
    })
}

#[cfg(feature = "noaa")]
fn cloud_cover_from_forecast(forecast: &str) -> i32 {
    let lower = forecast.to_lowercase();

    // Precipitation always means heavy cloud
    if lower.contains("rain")
        || lower.contains("storm")
        || lower.contains("snow")
        || lower.contains("drizzle")
        || lower.contains("showers")
    {
        return 95;
    }

    if lower.contains("overcast") {
        return 90;
    }

    // Mostly cloudy (before general "cloudy" check)
    if lower.contains("mostly cloudy") {
        return 75;
    }

    if lower.contains("cloudy") {
        return 90;
    }

    if lower.contains("partly") {
        return 50;
    }

    // Mostly sunny/clear (before general "sunny"/"clear")
    if lower.contains("mostly sunny") || lower.contains("mostly clear") {
        return 25;
    }

    if lower.contains("sunny") || lower.contains("clear") {
        return 10;
    }

    0
}

// Non-NOAA stubs
#[cfg(not(feature = "noaa"))]
pub fn init() {}

#[cfg(not(feature = "noaa"))]
pub fn cleanup() {}

#[cfg(not(feature = "noaa"))]
pub fn fetch(_lat: f64, _lon: f64) -> WeatherData {
    WeatherData {
        cloud_cover: 0,
        forecast: "Disabled (non-USA build)".to_string(),
        temperature: 0.0,
        is_day: true,
        fetched_at: now_epoch(),
        has_error: true,
    }
}
